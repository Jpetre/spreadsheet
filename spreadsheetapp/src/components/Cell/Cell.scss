.Cell {
  display: flex;
  align-items: center;
  width: 100px;
  padding: 5px;
  height: 40px;
  color: black;
  border-top: 1px solid #cacaca;
  border-left: 1px solid #cacaca;
  font-size: 15px;
  overflow: hidden;
  font-family: Calibri, Arial, Verdana, sans-serif;

  &--info {
    justify-content: center;
    background-color: #f0f0f0;
    font-weight: bold;
  }

  &--selected {
    outline-color: lightgrey;
    outline-style: dotted;
  }

  .Row:last-child & {
    border-bottom: 1px solid #cacaca;
  }

  &:last-child {
    border-right: 1px solid #cacaca;
  }

  @media screen and (max-width: 700px) {
    & {
      width: 60px;
      padding: 4px;
      height: 25px;
      font-size: 13px;
    }
  }
}
