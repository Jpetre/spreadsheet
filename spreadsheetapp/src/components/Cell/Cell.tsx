import React, { useState, useContext } from 'react';
import { SpreadSheetContext } from '../Table/Table';
import { CellType, CellDataType, SpreadSheetContextType } from '../../../../types/types';

interface CellProps extends CellType { }

const Cell: React.FC<CellProps> = (props: CellProps) => {
  const context: SpreadSheetContextType = useContext(SpreadSheetContext);
  const [editing, setEditing] = useState(false);
  const [inputValue, setInputValue] = useState(props.x > 0 && props.y > 0 ? context.data[props.x][props.y].inputValue : '');

  const updateContext = (value: string) => {
    const updatedData: { [key: string]: CellDataType } = context.data;
    updatedData[props.x][props.y].inputValue = value;
    updatedData[props.x][props.y].value = value;
    context.updateData(updatedData);
  };

  const handleClick = () => {
    setEditing(true);
  };

  const onBlur = (e: React.ChangeEvent<HTMLInputElement>) => {
    setEditing(false);
    updateContext(e.target.value);
  };

  // Cant type correctly to have both e.key and e.target.value working (with React.KeyboardEvent<HTMLInputElement>)
  const onKeyPressOnInput = (e: any) => {
    if (e.key === 'Enter') {
      setEditing(false);
      updateContext(e.target.value);
    };
  };

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  // header row
  if (props.x === 0) {
    return (
      <div className='Cell Cell--info'>
        {props.y}
      </div>
    )
  };
  // header column
  if (props.y === 0) {
    const alphabet = ' abcdefghijklmnopqrstuvwxyz'.split('');
    return (
      <div
        className='Cell Cell--info'>
        {alphabet[props.x]}
      </div>
    )
  };

  if (editing) {
    return (
      <input
        className='Cell'
        type='text'
        value={inputValue}
        onBlur={onBlur}
        onKeyPress={onKeyPressOnInput}
        onChange={onChange}
        autoFocus
      />
    )
  }

  return (
    <div
      className='Cell'
      onClick={() => handleClick()}
    >
      {context.data[props.x][props.y].value}
    </div>
  )
}

export default Cell;