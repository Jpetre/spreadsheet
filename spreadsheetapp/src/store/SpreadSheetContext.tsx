import React from 'react';
import { ColumnDataType } from '../../../types/types';

export const SpreadSheetContext = React.createContext({
  data: {},
  updateData: (data: ColumnDataType) => { }
});
