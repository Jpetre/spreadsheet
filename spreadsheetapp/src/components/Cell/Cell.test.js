import React from 'react';
import Cell from './Cell';
import renderer from 'react-test-renderer';
import { SpreadSheetContext } from '../../store/SpreadSheetContext';
import mockData from '../../mocks/data.json';

test('Cell render correctly', () => {
  const component = renderer.create(
    <SpreadSheetContext.Provider
      value={{
        data: mockData
      }}
    >
      <Cell 
        y={1}
        x={1}
      />
    </SpreadSheetContext.Provider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Cell header column render correctly', () => {
  const component = renderer.create(
    <Cell 
      y={0}
      x={1}
    />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Cell header row render correctly', () => {
  const component = renderer.create(
    <Cell 
      y={1}
      x={0}
    />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});