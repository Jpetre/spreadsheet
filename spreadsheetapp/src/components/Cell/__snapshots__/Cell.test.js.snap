// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell header column render correctly 1`] = `
<div
  className="Cell Cell--info"
>
  a
</div>
`;

exports[`Cell header row render correctly 1`] = `
<div
  className="Cell Cell--info"
>
  1
</div>
`;

exports[`Cell render correctly 1`] = `
<div
  className="Cell"
  onClick={[Function]}
>
  12
</div>
`;
